docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,608 @@
+Child devices follow a similar structure for commands as well:
+
+```
+tedge/<child-id>/commands/req/software_list
+tedge/<child-id>/commands/res/software_update
+```
+
+
+If command support is added to services in future, they'd follow a similar pattern as in:
+
+```
+tedge/<device-id>/service/<service-id>/commands/req/software_list
+tedge/<device-id>/service/<service-id>/commands/req/software_list
+```
+
+**Examples:**
+
+* Software list operation
+
+   ```
+   tedge/main/device/commands/req/software_list
+   tedge/main/device/commands/res/software_list
+   ```
+
+* Software update operation
+
+   ```
+   tedge/<child-id>/device/commands/req/software_update
+   tedge/<child-id>/device/commands/res/software_update
+   ```
+
+* Firmware update operation
+
+   ```
+   tedge/main/device/commands/req/firmware_update
+   tedge/main/device/commands/res/firmware_update
+   ```
+
+* Device restart operation
+
+   ```
+   tedge/main/device/commands/req/device_restart
+   tedge/main/device/commands/res/device_restart
+   ```
+
+* Configuration snapshot operation
+
+   ```
+   tedge/main/service/<service-id>/commands/req/config_snapshot
+   tedge/main/service/<service-id>/commands/res/config_snapshot
+   ```
+
+* Configuration update operation
+
+   ```
+   tedge/<child-id>/service/<service-id>/commands/req/config_update
+   tedge/<child-id>/service/<service-id>/commands/res/config_update
+   ```
+
+Although all the above examples maintain consistent structure by ending with the `<operation-type>`,
+further additions are possible in future if desired for a given operation type.
+For e.g: `tedge/main/commands/req/config_update/<config-type>` to address a specific `config-type`
+
+### Registration service for child devices
+
+Immediate and nested child devices can be registered with thin-edge using its registration service,
+by sending the following MQTT message to the topic: `tedge/main/register/req/child`:
+
+```json
+{ 
+   "parent": "<parent-device-id>",
+   "id-prefix": "<desired-child-id-prefix>",
+   "capabilities": {
+      "<capability-1>": {},  //capability-1 specific metadata
+      "<capability-2>": {},  //capability-2 specific metadata
+      ...
+   }
+}
+```
+
+The `parent-device-id` is the device-id of the direct parent that the child device is connected to.
+The payload can have other fields describing the capabilities of that device as well (config management, software management etc).
+
+```admonish warning
+The exact topic keys and payload format for this init contract can be discussed and refined separately.
+The focus here is just on the MQTT topic structure.
+```
+
+Thin-edge needs to maintain the lineage (hierarchy of parent devices) of all descendent child devices in its internal state,
+so that it can be looked up while receiving any data from them.
+Even though the child device only declares its immediate parent in the registration message,
+the entire lineage can be traced back with a recursive lookup on that `parent-device-id`.
```

Comment:
Okay, but with the assumption the parent is already registered.

Reply:
Yes. The registration would fail if a child tries to register with a non-existent parent. Yes, this means that this ordering (child only after parent) must be respected when an entire fleet is bootstrapped together.